### PR TITLE
OpenAPI No inheritance in connector_mgmt spec

### PR DESCRIPTION
**Pull request by andreaTP**
## Description
Following up on the work started here: #1518
And sneaking in a minor workaround that seems to not be affecting the generated code at all.

cc. @lburgazzoli 

## Checklist (Definition of Done)

- [ ] All acceptance criteria specified in JIRA have been completed
- [ ] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] Documentation added for the feature
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] Required metrics/dashboards/alerts have been added (or PR created).
- [ ] Required Standard Operating Procedure (SOP) is added.
- [ ] JIRA has been created for changes required on the client side

### CODE REVIEW

File: openapi/connector_mgmt.yaml [hunk 1]
```diff
@@ -1122,9 +1117,8 @@ components:
           properties:
             collections:
               type: array
-              items:
-                allOf:
-                  - $ref: "#/components/schemas/ObjectReference"
+              items: # removed allOf to workaround this issue: https://github.com/microsoft/kiota/issues/2442
+                $ref: "#/components/schemas/ObjectReference"
```
Comment (andreaTP) [inline openapi/connector_mgmt.yaml:1121]: this is the workaround I'm sneaking in, the generated code is not affected and the signature is pretty much compatible with the previous; I do believe that it's an acceptable change to make the life of the SDKs generator easier 🙂

**Comment by andreaTP** [comment]
The CI failure seems to be unrelated to the changes:

> Codecov: Failed to properly upload: The process '/home/runner/work/_actions/codecov/codecov-action/v3/dist/codecov' failed with exit code 255

**Comment by manstis** [comment]
This leaves a `List` model in the public schema that (1) is not used, (2) lacks `items` (and hence arguably is not a list).

IDK whether it's better to dispense with the common `List` all together and embed the necessary properties directly. e.g.
```
    ConnectorClusterList:
      required:
        - kind
        - page
        - size
        - total
        - items
      allOf:
        - type: object
          properties:
            kind:
              type: string
            page:
              type: integer
            size:
              type: integer
            total:
              type: integer
            items:
              type: array
              items:
                $ref: "#/components/schemas/ConnectorCluster"
``` 
This generates the same Go model structures and removes the surplus `List` model.

WDYT @lburgazzoli @valdar ?

**Comment by andreaTP** [comment]
@manstis I have gone down the path of doing minimal changes to achieve the result, I don't have a strong opinion on the point you rise, but I have to say that the proposed PR is modeled after the rest of the APIs we are exposing.

File: openapi/kas-fleet-manager.yaml [hunk 2]
```diff
@@ -1571,9 +1571,8 @@ components:
             type: string
           collections:
             type: array
-            items:
-              allOf:
-              - $ref: "#/components/schemas/ObjectReference"
+            items: # removed allOf to workaround this issue: https://github.com/microsoft/kiota/issues/2442
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply openapi/kas-fleet-manager.yaml:1574]: just this specific case is triggering a bug

**Comment by machi1990** [comment]
> This leaves a `List` model in the public schema that (1) is not used, (2) lacks `items` (and hence arguably is not a list).
> 
> IDK whether it's better to dispense with the common `List` all together and embed the necessary properties directly. e.g.
> 
> ```
>     ConnectorClusterList:
>       required:
>         - kind
>         - page
>         - size
>         - total
>         - items
>       allOf:
>         - type: object
>           properties:
>             kind:
>               type: string
>             page:
>               type: integer
>             size:
>               type: integer
>             total:
>               type: integer
>             items:
>               type: array
>               items:
>                 $ref: "#/components/schemas/ConnectorCluster"
> ```
> 
> This generates the same Go model structures and removes the surplus `List` model.
> 
> WDYT @lburgazzoli @valdar ?

@bf2fc6cc711aee1a0c2a/kas-fleet-manager-reviewers any concern regarding this?

File: openapi/kas-fleet-manager.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (miguelsorianod) [reply openapi/kas-fleet-manager.yaml:1574]: Does this change the format of the response? I am thinking if this has any impact on the API. Or it is just related to the Go autogeneration.
Has `make generate` also been run to verify no changes at Go code level happen?

Thanks

[the remark this answers is not in the document]
Comment (andreaTP) [reply openapi/kas-fleet-manager.yaml:1574]: I'm following the instructions here:
https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/main/CONTRIBUTING.md#modifying-the-api-definition

and yes, this PR include the code generated by `make openapi/generate`.

[the remark this answers is not in the document]
Comment (miguelsorianod) [reply openapi/kas-fleet-manager.yaml:1574]: :+1: . As long as it doesn't change the format of the response returned I am fine with it. The only concern is/was API compatibility on the responses

**Comment by miguelsorianod** [comment]
> > This leaves a `List` model in the public schema that (1) is not used, (2) lacks `items` (and hence arguably is not a list).
> > IDK whether it's better to dispense with the common `List` all together and embed the necessary properties directly. e.g.
> > ```
> >     ConnectorClusterList:
> >       required:
> >         - kind
> >         - page
> >         - size
> >         - total
> >         - items
> >       allOf:
> >         - type: object
> >           properties:
> >             kind:
> >               type: string
> >             page:
> >               type: integer
> >             size:
> >               type: integer
> >             total:
> >               type: integer
> >             items:
> >               type: array
> >               items:
> >                 $ref: "#/components/schemas/ConnectorCluster"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This generates the same Go model structures and removes the surplus `List` model.
> > WDYT @lburgazzoli @valdar ?
> 
> @bf2fc6cc711aee1a0c2a/kas-fleet-manager-reviewers any concern regarding this?

I agree with Michael. The `List` type should contain the `items` property as it is the standardized API response for list of results.
If part of your response, which is not a `List` (does not contain the `items` attribute) contains some attribute that is an *array* (an ordered set of elements) then it's not a `List`

**Comment by andreaTP** [comment]
> I agree with Michael. The List type should contain the items property as it is the standardized API response for list of results.

The issue we are solving here is the fact that JsonSchema doesn't allow overwriting fields with name clashes, so `items` **should** be excluded from `List` as it cannot be "extended overwriting" it(the concept itself doesn't exist in JsonSchema).

If you want to remove the `List` model at all, I personally don't mind, but I would like to ask if we can separate this discussion from this PR as, whatever the outcome, it can be implemented on top of this PR without breaking changes.

**Comment by manstis** [comment]
> If you want to remove the `List` model...

FWIW I don't have a **strong** opinion. 

The purist in me says leaving a structure called `List` that actually lacks any items makes it a non-List.

The pragmatist in me says move on; it's not a big deal.

**Comment by lburgazzoli** [comment]
As long as it does not break the generated code, then any solution is fine with me.

Maybe in the future we can have a `ListMeta` so  concrete lists would be something like:

```yaml
    ConnectorClusterList:
      allOf:
        - $ref: "#/components/schemas/ListMeta"
          properties:
        - type: object
          properties:
            items:
              type: array
              items:
                $ref: "#/components/schemas/ConnectorCluster"
```

Note: I don't know if that's valid :)

**Review by machi1990 — APPROVED** [review]

**Review by lburgazzoli — APPROVED** [review]

**Comment by miguelsorianod** [comment]
> > I agree with Michael. The List type should contain the items property as it is the standardized API response for list of results.
> 
> The issue we are solving here is the fact that JsonSchema doesn't allow overwriting fields with name clashes, so `items` **should** be excluded from `List` as it cannot be "extended overwriting" it(the concept itself doesn't exist in JsonSchema).
> 
> If you want to remove the `List` model at all, I personally don't mind, but I would like to ask if we can separate this discussion from this PR as, whatever the outcome, it can be implemented on top of this PR without breaking changes.

I see what you mean now. That is the correct approach.
I was reviewing what we did in the KAS Fleet Manager OpenAPI types and I remembered we also came up with that. In that case we have a `List` type that does not have the `items` property, the type being equivalent to the `ListMeta` that Luca was proposing and exactly what's being done in this PR. Then, in the `*List` types we include the `List` and add the `items` property as required as being shown in this PR. You can see an example with KafkaRequestList.

I am fine with the proposed approach in the PR, and if you want to rename `List` to `ListMeta` too in a future PR too of course. At that point is just a semantic change

**Review by miguelsorianod — APPROVED** [review]

**Comment by lburgazzoli** [comment]
> `ListMeta` that Luca was proposing and exactly what's being done in this PR. 

I got confused by all the component and oldness :)

**Pull request merged**